### ISSUE TITLE
Getting error when hitting checkout/payment from checkout/address

### ISSUE THREAD

**Issue by twd3** [duncanmcclean/simple-commerce#307]
I edited the checkout process a bit. Got rid of account and shipping. In my setup they go:
login>add to cart>cart>checkout/email>checkout/address>checkout/payment

When I hit payment I'm getting:
**Argument 1 passed to DoubleThreeDigital\SimpleCommerce\Repositories\CustomerRepository::findByEmail() must be of the type string, null given, called in C:\_statamic\site\vendor\laravel\framework\src\Illuminate\Support\Facades\Facade.php on line 261** 

I checked to see if there is a customer entry for this test user and there is. The file is named: test-at-testnet.md I assume that is right. I was going to get rid of the checkout/email step too because they're logged in but I think the cart needs it?

I'm a bit worried about this one though because I've seen the payment step work in this setup. I'm just using the default checkout.

...

I expected to see the stripe payment with a cart to the right.

...

## Environment

Statamic 3.0.13 Pro
Laravel 7.26.1
PHP 7.4.6
doublethreedigital/simple-commerce 2.0.17
optimoapps/statamic-bard-text-align 1.0.0
spatie/statamic-responsive-images 1.2.3

**Comment by duncanmcclean** [comment]
A few questions:

* Is there a `customer` set on the order entry, you should be able to see it in the file. It should have the ID of the customer entry.

* Can you share the full stack trace so I can see where it is that's calling the `CustomerRepository`?

**Comment by twd3** [comment]
Okay, no, after I hit the email step it adds the email to the order but then I hit address and it doesn't add anything to the order, then I hit payment and get this error.

Here is what the order .md files ends up looking like:
```
---
title: '#2001'
items:
  -
    id: 87e0072a-410c-4c5f-929a-3a958a40d022
    product: 96a4e71d-6384-495a-93d3-815cd904350e
    quantity: 1
    total: 100500
is_paid: false
grand_total: 100701
items_total: 100500
tax_total: 201
shipping_total: 0
coupon_total: 0
order_status: cart
email: test@test.net
id: de3c35fb-a731-413d-ba00-be50f1577229
published: false
---
```
Here is that customer (customer is also a user):
```
---
email: test@test.net
id: 03f192e7-8df7-4f25-a8c0-046d672c87c9
published: false
---
```

**Comment by duncanmcclean** [comment]
Hmm, I'll have a look in a bit... Seems like something strange is going on.

**Comment by duncanmcclean** [comment]
I've fixed the bug you were having and added some tests to ensure this issue doesn't happen again.

I've created a pull request for the fix, #308 which I'll review and release in the morning when I'm more awake 😂

**Comment by duncanmcclean** [comment]
Released a bugfix in [v2.0.18](https://github.com/doublethreedigital/simple-commerce/releases/tag/v2.0.18), could you confirm the fix works?

**Comment by twd3** [comment]
Thanks Duncan! I Updated and tried again. Now when I hit payment from address I get: https://flareapp.io/share/VP6K3V37

**Comment by twd3** [comment]
Thank Duncan! Okay, after clearing old test orders for customer I see it does add the name and address. But now when I hit payment I can only see the cart, there is no stripe payment. Here is my payment part of check out. It's default except for I am hiding the "payment method" drop because it's always just stripe...
```
    {{ sc:checkout redirect="/checkout/complete" class="flex-1 mb-8 md:mb-0" }}
        <div x-data="{ gateway: '{{ sc:gateways }}{{ if first }}{{ formatted_class }}{{ /if }}{{ /sc:gateways }}' }">
            <div class="hidden w-full md:w-1/2 mb-4 flex flex-col">
                <label
                    for="payment_method"
                    class="text-gray-800 font-semibold mb-1"
                >Payment Method</label>

                <div class="inline-block relative w-full p-2 bg-gray-200">
                    <select
                        x-model="gateway"
                        class="block appearance-none w-full bg-gray-200 text-gray-800 focus:outline-none"
                        id="payment_method"
                        name="gateway"
                        value="{{ old:gateway }}"
                        required
                    >
                        <option disabled>Choose a Payment Method</option>
                        {{ sc:gateways }}
                            <option value="{{ class }}">{{ name }}</option>
                        {{ /sc:gateways }}
                    </select>

                    <div class="pointer-events-none absolute inset-y-0 right-0 flex items-center px-2 text-gray-700">
                        <svg class="fill-current h-4 w-4" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20"><path d="M9.293 12.95l.707.707L15.657 8l-1.414-1.414L10 10.828 5.757 6.586 4.343 8z"/></svg>
                    </div>
                </div>
            </div>

            {{ sc:gateways }}
                <div class="w-full" x-show="gateway === '{{ formatted_class }}'">
                    {{ partial src="checkout/gateways/{ handle }" }}
                </div>
            {{ /sc:gateways }}

            <button
                class="bg-blue-500 text-white font-bold uppercase flex items-center justify-center px-6 py-2"
                type="submit"
                @click.prevent="loading = true; if(typeof confirmPayment == 'function' && document.getElementsByName('gateway')[0].value == 'DoubleThreeDigital\\SimpleCommerce\\Gateways\\StripeGateway') { confirmPayment(); } else { document.querySelector('form').submit() }"
            >Pay {{ sc:cart }}{{ grand_total }}{{ /sc:cart }}</button>
        </div>
    {{ /sc:checkout }}
```

**Comment by duncanmcclean** [comment]
I'll have a look later, I'm not at my computer right now.

**Comment by duncanmcclean** [comment]
I'm back at my computer now - could you try commenting out lines 41 to 50 in your StripeGateway.php file, which will be located in `vendor/doublethreedigital/simple-commerce/src/Gateways/StripeGateway.php`. [(see lines in question)](https://github.com/doublethreedigital/simple-commerce/blob/master/src/Gateways/StripeGateway.php#L41)

Those lines are to do with passing the customer into Stripe, I wonder if something's wrong with that and that's why the Stripe form is throwing an exception behind the scenes, that causes the form to not display at all.

Let me know how you get on with that.... 

(If you'd prefer to chat on Discord that may be a better option to avoid long delays between replies 🙂)

**Comment by twd3** [comment]
Commenting out lines fixed my issue.

**Comment by duncanmcclean** [comment]
Sweet. I'll get that bit fixed up tomorrow.

**Comment by duncanmcclean** [comment]
Alright.... just pushed out what is hopefully a bug fix to the StripeGateway issue... [v2.0.19](https://github.com/doublethreedigital/simple-commerce/releases/tag/v2.0.19).

Hopefully that fixes the problem.

**Comment by twd3** [comment]
I just tested after updating and got the error below. Maybe because shipping_name is on address step not just name? IDK...

**Comment by duncanmcclean** [comment]
Looks like that error is coming from when it tries to generate a title for the customer entry. Looks like, for whatever reason, the customer entry didn't have a `name` field filled in which is what must be causing this issue (could be potentially what caused the Stripe gateway issue as well).

Basically, Simple Commerce wasn't checking a name field was even set before trying to include it in the title. I'm going to push out a bug fix shortly that'll check that so we don't run into the same thing again. I appreciate you hanging along while I fix these bugs 😅

**Comment by twd3** [comment]
Yeah, I just tried adding a name input to the email step so the order and customer got a name and still got Undefined index: name.

**Comment by duncanmcclean** [comment]
Here's the release I was talking about: [v2.0.20](https://github.com/doublethreedigital/simple-commerce/releases/tag/v2.0.20)

**Comment by duncanmcclean** [comment]
> Yeah, I just tried adding a name input to the email step so the order and customer got a name and still got Undefined index: name.

Where did you add the customer name field? Inside a `{{ sc:cart:update }}` form?

**Comment by twd3** [comment]
Onward and upward. Yup:
```
{{ sc:cart:update redirect="/checkout/address" }}
	<div class="w-full md:w-1/3 mb-4 flex flex-col">
		
		<label
			for="name"
			class="text-gray-800 font-semibold mb-1"
		>Name</label>

		<input
			class="bg-gray-200 p-2 text-gray-800"
			id="name"
			type="text"
			name="name"
			placeholder="Full Name"
			value="{{ user }}{{ name }}{{ /user }}"
			required
		>
		
		<label
			for="email"
			class="text-gray-800 font-semibold mb-1"
		>Email</label>

		<input
			class="bg-gray-200 p-2 text-gray-800"
			id="email"
			type="email"
			name="email"
			placeholder="you@example.com"
			value="{{ user }}{{ email }}{{ /user }}"
			required
		>
	</div>

	<button
		class="bg-blue-500 text-white font-bold uppercase flex items-center justify-center px-6 py-2"
		type="submit"
	>Continue</button>
{{ /sc:cart:update }}
```

**Comment by duncanmcclean** [comment]
Try changing the name of the name and email inputs to `customer[name]` and `customer[email]` instead and see if that adds the name & email to the customer.

**Comment by twd3** [comment]
On it. I'll update first real quick.

**Comment by twd3** [comment]
https://flareapp.io/share/LPlWexgm

**Comment by twd3** [comment]
Okay, I just deleted the customer and tried again and got different error: Typed property DoubleThreeDigital\SimpleCommerce\Repositories\CustomerRepository::$title must be string, null used 
After deleting customer and giving it a go this is what it made for the customer file and it gave error above on payment step.
```
---
name: 'Test User'
email: test@test.net
id: fe95d7b3-e548-48ba-be96-8498433ef7c7
published: false
---
```

**Comment by duncanmcclean** [comment]
Hmm, could you share the Flare stack trace for that one as well please?

**Comment by twd3** [comment]
Trying to it won't. Hold on I'll try in reg window and not inprivate.

**Comment by twd3** [comment]
Still says unable to share...

**Comment by duncanmcclean** [comment]
Okay... hopefully I've come up with a fix for that error....

Could you comment out/remove [lines 39 to 45](https://github.com/doublethreedigital/simple-commerce/blob/master/src/Repositories/CustomerRepository.php#L39) of `vendor/doublethreedigital/simple-commerce/src/Repositories/CustomerRepository.php` and put the following in it's place?

```php
if ($entry->title === null || $entry->entry === '' || $entry->slug === null || $entry->slug === '') {
  $this->generateTitleAndSlug();
} else {
  $this->title = $entry->title;
  $this->slug = $entry->slug;
}
```

It basically needs to check if the title and the slug are null, instead of checking if they are empty strings, like it does right now.

**Comment by twd3** [comment]
Yup. Hold on.

**Comment by twd3** [comment]
I think you got it! Went all the way through to Stripe. Made customer and tied customer to successful payment.

**Comment by duncanmcclean** [comment]
Whoop whoop! I'll release a patch for that now I know it fixes the problem 😅

**Comment by twd3** [comment]
Nice! TY

**Comment by duncanmcclean** [comment]
No problem!

**Comment by duncanmcclean** [comment]
Fixed in [v2.0.21](https://github.com/doublethreedigital/simple-commerce/releases/tag/v2.0.21)
